### PR TITLE
Add CREFL rayleigh corrector for ABI data

### PR DESCRIPTION
**Pull request by wroberts4**
I added the rayleigh correction to also work for abi data in order to incorporate a wider range of satpy functionality. The code I wrote was re-purposed from work done by Ralph Kuehn at SSEC.

Note: Ralph's code uses a different algorithm than that which is used by satpy.  A research paper covering the algorithm that Ralph's code uses can be found here: https://www.atmos-meas-tech.net/6/2989/2013/

A couple of things to keep in mind:
1. Added modifiers in abi/viirs.yaml may need be re-purposed/removed in the future. Same goes for composites in abi.yaml.
2. The last 5 elements of bH2O in crefl_utils.py are actually aO2 values (in Ralph's code). Currently aO2 is not used in the satpy code since it would require restructuring of multiple features, so they are tagged at the end of bH2O.
 - [x] Tests added 
 - [x] Tests passed 
 - [x] Passes ``git diff origin/master **/*py | flake8 --diff`` 
 - [x] Fully documented

### CODE REVIEW

**Comment by djhoese** [comment]
If I recall correctly @wroberts4 this still produces different results that Ralph's original code. Maybe you could try running the same data through this satpy code and Ralph's old code and show us screenshots of the results? Or I guess screenshots once they are equivalent.

Then just unittests to get the coverage score higher and maybe @adybbroe and @mraspaud would agree that this should maybe be moved to `composites/__init__.py`? Or maybe make the full jump to adding it to pyspectral?

**Comment by mraspaud** [comment]
About moving to `composites/__init__.py` I don't know, I was just looking at the file the other day, and thought it was starting to get big. Should we maybe start splitting it instead ?

**Comment by djhoese** [comment]
@mraspaud Sounds good to me. Wondering if we should have `_rayleigh.py` or maybe a `atmospheric_correctors.py` or something like that?

**Comment by mraspaud** [comment]
`atmospheric_correctors.py` sounds good

File: satpy/composites/viirs.py [hunk 1]
```diff
@@ -38,6 +39,14 @@
 LOG = logging.getLogger(__name__)
 
 
+class IncompatibleAreas(Exception):
```
Comment (djhoese) [inline satpy/composites/viirs.py]: This should have been imported from `composites/__init__.py`

File: satpy/composites/crefl_utils.py [hunk 2]
```diff
@@ -331,6 +363,74 @@ def get_atm_variables(mus, muv, phi, height, coeffs):
...
+    xitot2 = xph2 * (xitm1 + xitm2 * fs1)
+    xitot3 = xph3 * (xitm1 + xitm2 * fs2)
+    rhoray = xitot1 * xcos1 + xitot2 * xcos2 * 2.0 + xitot3 * xcos3 * 2.0
+
+    if isinstance(height, xr.DataArray):
+        def _sphalb_index(sphalb0, index_arr):
+            # FIXME: if/when dask can support lazy index arrays then remove this
+            return sphalb0[index_arr]
+        sphalb_delayed = dask.delayed(_sphalb_index)(sphalb0, (taur / TAUSTEP4SPHALB + 0.5).astype(np.int32))
+        sphalb = da.from_delayed(sphalb_delayed, taur.shape, dtype=sphalb0.dtype)
```
Comment (djhoese) [inline satpy/composites/crefl_utils.py]: I thought about this while working on a separate feature. I think you could get this to work with `map_blocks` if you provide `index_arr` as the first argument, specify the `chunks=taur.chunks`, and most importantly before passing the `sphalb0` array do `sphalb0 = sphalb0.rechunk(sphalb0.shape)`. We can talk about this in person later.

**Comment by djhoese** [comment]
I think maybe the crefl scaling should be renamed to "contrast_curves". @pnuu didn't we discuss this? What was the best name we came up with?

File: satpy/composites/crefl_utils.py [hunk 3]
```diff
@@ -359,44 +459,53 @@ def run_crefl(refl, coeffs,
...
-        tOG = np.repeat(np.repeat(tOG, factor, axis=0), factor, axis=1)
-        TtotraytH2O = np.repeat(
-            np.repeat(TtotraytH2O, factor, axis=0),
-            factor, axis=1)
-        # if average height wasn't provided then this should stay a scalar
-        if sphalb.size != 1:
-            # otherwise make it the same size as the other arrays
-            sphalb = np.repeat(np.repeat(sphalb, factor, axis=0), factor, axis=1)
+        from satpy.composites import IncompatibleAreas
+        raise IncompatibleAreas("")
```
Comment (djhoese) [inline satpy/composites/crefl_utils.py]: I would say remove this entire if block. This should theoretically not be possible when called from the modifier (because of `check_areas`).

File: satpy/composites/viirs.py [hunk 4]
```diff
@@ -85,18 +86,37 @@ def __init__(self, *args, **kwargs):
...
         if os.path.isfile(self.dem_file):
             LOG.debug("Loading CREFL averaged elevation information from: %s",
                       self.dem_file)
             from netCDF4 import Dataset as NCDataset
             # HDF4 file, NetCDF library needs to be compiled with HDF4 support
             nc = NCDataset(self.dem_file, "r")
-            avg_elevation = nc.variables[self.dem_sds][:]
+            # average elevation is stored as a 16-bit signed integer but with
+            # scale factor 1 and offset 0, convert it to float here
+            avg_elevation = nc.variables[self.dem_sds][:].astype(np.float)
```
Comment (djhoese) [inline satpy/composites/viirs.py]: I think in the long run we'll want this to be dask arrays, but for now let's not worry about it. It probably isn't a performance issue anyway.

File: satpy/enhancements/__init__.py [hunk 5]
```diff
@@ -110,6 +110,23 @@ def apply_enhancement(data, func, exclude=None, separate=False,
     return data
 
 
+# pointed to by generic.yaml
+def crefl_scaling(img, **kwargs):
```
Comment (djhoese) [inline satpy/enhancements/__init__.py]: As discussed on slack I think we'll have to make another issue to combine this with the existing crude stretch so that it allows for multiple "control" points.

File: satpy/etc/composites/abi.yaml [hunk 6]
```diff
@@ -1,6 +1,15 @@
 sensor_name: visir/abi
 
 modifiers:
+  # TODO: fix
```
Comment (djhoese) [inline satpy/etc/composites/abi.yaml]: Remove `# TODO`. Also future note, usually a todo of fixing something is just `# FIXME: description` 😉

File: satpy/etc/enhancements/generic.yaml [hunk 7]
```diff
@@ -198,7 +207,7 @@ enhancements:
     operations:
     - name: stretch
       method: *stretchfun
-      kwargs: {stretch: linear}
+      kwargs: {stretch: crude, min_stretch: 0, max_stretch: 120}
```
Comment (djhoese) [inline satpy/etc/enhancements/generic.yaml]: @mraspaud @adybbroe @pnuu Thoughts on changing the default reflectance scaling to a static 0 to 120?

File: satpy/node.py [hunk 8]
```diff
@@ -423,12 +423,10 @@ def _find_dependencies(self, dataset_key, **dfilter):
         except KeyError:
             # exact dataset isn't loaded, let's load it below
             pass
-
```
Comment (djhoese) [inline satpy/node.py]: Why were these lines removed? Automatic by some auto-correcting tool? Please add the blank lines back in unless they were style errors.

File: satpy/tests/test_crefl_utils.py [hunk 9]
```diff
@@ -0,0 +1,30 @@
+import unittest
+
+
+class testCreflUtils(unittest.TestCase):
```
Comment (djhoese) [inline satpy/tests/test_crefl_utils.py]: Since your my student @wroberts4 I'm going to ask you to fix the name of this class and the name of your test functions to be PEP8 compatible. `TestCreflUtils` and `test_get_atm_variables_abi`

Comment (djhoese) [reply satpy/tests/test_crefl_utils.py]: Try to review other tests that you've added.

File: satpy/composites/crefl_utils.py [hunk 10]
```diff
@@ -331,6 +363,73 @@ def get_atm_variables(mus, muv, phi, height, coeffs):
     return sphalb, rhoray, TtotraytH2O, tOG
 
 
+def get_atm_variables_abi(mus, muv, phi, height, coeffs, G_O3, G_H2O, G_O2):
```
Comment (mraspaud) [inline satpy/composites/crefl_utils.py]: there seems to be a lot of duplicated code in this function from existing function, can we factorise ?

Comment (wroberts4) [reply satpy/composites/crefl_utils.py]: You are correct! I cooked something up and will push it later today or Monday (after I test to make sure I didn't break it).

File: satpy/etc/readers/hdfeos_l1b.yaml [hunk 11]
```diff
@@ -21,7 +21,9 @@ datasets:
       250: {file_type: hdf_eos_data_250m}
       500: {file_type: hdf_eos_data_500m}
       1000: {file_type: hdf_eos_data_1000m}
-    calibration: reflectance
+    calibration:
+      reflectance:
+        units: "%"
```
Comment (djhoese) [inline satpy/etc/readers/hdfeos_l1b.yaml]: Do we know that the units coming out of this are `%` (0-~120) and not `1` (0 - ~1.2)?

File: satpy/etc/composites/modis.yaml [hunk 12]
```diff
@@ -12,15 +21,15 @@ composites:
       modifiers: [sunz_corrected]
     standard_name: true_color
 
-  true_color:
+  true_color_crefl:
```
Comment (djhoese) [inline satpy/etc/composites/modis.yaml]: Please reset this to be `true_color` and to use `rayleigh_corrected` instead of `rayleigh_corrected_crefl`. Feel free to leave the `rayleigh_corrected_crefl` definition above. Make sure that this true color composite still loads successfully.

File: satpy/tests/test_crefl_utils.py [hunk 13]
```diff
@@ -0,0 +1,30 @@
...
+
+    def test_get_atm_variables_abi(self):
+        import numpy as np
+        from satpy.composites.crefl_utils import get_atm_variables_abi
+        sphalb, rhoray, TtotraytH2O, tOG = get_atm_variables_abi(0.17690244, 6.123234e-17, 530.61332168, 405.,
+                                                                 21.71342113, 77.14385758, 56.214566960,
+                                                                 0.0043149700000000004, 0.0037296,
+                                                                 0.014107995000000002, 0.052349)
+        if abs(np.array(sphalb) - 0.045213532544630494) >= 1e-10:
+            raise AssertionError('{} is not within {} of {}'.format(np.array(sphalb), 1e-10, 0.045213532544630494))
```
Comment (djhoese) [inline satpy/tests/test_crefl_utils.py]: Sorry, I should have noticed these earlier. These should be replaced with assert statements or `self.assertX` calls.

File: satpy/tests/compositor_tests/test_viirs.py [hunk 14]
```diff
@@ -223,6 +242,254 @@ def test_hncc_dnb(self):
...
+
+        self.assertEqual(ref_cor.attrs['modifiers'], ('sunz_corrected', 'rayleigh_corrected_crefl',))
+        self.assertEqual(ref_cor.attrs['calibration'], 'reflectance')
+        self.assertEqual(ref_cor.attrs['wavelength'], (0.45, 0.47, 0.49))
+        self.assertEqual(ref_cor.attrs['name'], 'C01')
+        self.assertEqual(ref_cor.attrs['resolution'], 1000)
+        self.assertEqual(ref_cor.attrs['sensor'], 'abi')
+        self.assertEqual(ref_cor.attrs['prerequisites'], [])
+        self.assertEqual(ref_cor.attrs['optional_prerequisites'], [
+            DatasetID(name='satellite_azimuth_angle', wavelength=None, resolution=None, polarization=None,
```
Comment (djhoese) [inline satpy/tests/compositor_tests/test_viirs.py]: The default for DatasetID is `None` so these can all be left empty (except for maybe `modifiers` which defaults to an empty tuple).

File: satpy/tests/compositor_tests/test_viirs.py [hunk 15]
```diff
@@ -223,6 +242,254 @@ def test_hncc_dnb(self):
...
+        self.assertEqual(res.attrs['name'], 'C01')
+        self.assertEqual(res.attrs['resolution'], 1000)
+        self.assertEqual(res.attrs['sensor'], 'abi')
+        self.assertEqual(res.attrs['start_time'], '2017-09-20 17:30:40.800000')
+        self.assertEqual(res.attrs['end_time'], '2017-09-20 17:41:17.500000')
+        self.assertEqual(res.attrs['area'], area)
+        self.assertEqual(res.attrs['ancillary_variables'], [])
+        data = res.values
+        if abs(np.mean(data) - 29.907390988422513) >= 1e-10:
+            raise AssertionError('{} is not within {} of {}'.format(np.mean(data), 1e-10, 29.907390988422513))
```
Comment (djhoese) [inline satpy/tests/compositor_tests/test_viirs.py]: Replace this with `self.assertX` call or regular assert statement at least.

File: satpy/tests/compositor_tests/test_viirs.py [hunk 16]
```diff
@@ -223,6 +242,254 @@ def test_hncc_dnb(self):
...
+
+    def test_reflectance_corrector_viirs(self):
+        import xarray as xr
+        import dask.array as da
+        import numpy as np
+        import datetime
+        from satpy.composites.viirs import ReflectanceCorrector
+        from satpy import DatasetID
+        ref_cor = ReflectanceCorrector(dem_filename='_fake.hdf', optional_prerequisites=[
+         DatasetID(name='satellite_azimuth_angle', wavelength=None, resolution=371, polarization=None, calibration=None,
```
Comment (djhoese) [inline satpy/tests/compositor_tests/test_viirs.py]: No need to list each `None` element here.

File: satpy/tests/compositor_tests/test_viirs.py [hunk 17]
```diff
@@ -223,6 +242,254 @@ def test_hncc_dnb(self):
...
+        self.assertEqual(ref_cor.attrs['modifiers'], ('sunz_corrected_iband', 'rayleigh_corrected_crefl_iband'))
+        self.assertEqual(ref_cor.attrs['calibration'], 'reflectance')
+        self.assertEqual(ref_cor.attrs['wavelength'], (0.6, 0.64, 0.68))
+        self.assertEqual(ref_cor.attrs['name'], 'I01')
+        self.assertEqual(ref_cor.attrs['resolution'], 371)
+        self.assertEqual(ref_cor.attrs['sensor'], 'viirs')
+        self.assertEqual(ref_cor.attrs['prerequisites'], [])
+        self.assertEqual(ref_cor.attrs['optional_prerequisites'], [
+            DatasetID(name='satellite_azimuth_angle', wavelength=None, resolution=371, polarization=None,
+                      calibration=None, level=None, modifiers=None),
```
Comment (djhoese) [inline satpy/tests/compositor_tests/test_viirs.py]: Same here.

File: satpy/tests/compositor_tests/test_viirs.py [hunk 18]
```diff
@@ -223,6 +242,254 @@ def test_hncc_dnb(self):
...
+            DatasetID(name='solar_azimuth_angle', wavelength=None, resolution=371, polarization=None, calibration=None,
+                      level=None, modifiers=None),
+            DatasetID(name='solar_zenith_angle', wavelength=None, resolution=371, polarization=None, calibration=None,
+                      level=None, modifiers=None)
+            ])
+
+        area, dnb = self.data_area_ref_corrector()
+
+        def make_xarray(wavelength, modifiers, calibration, file_type, file_key, name, standard_name, units):
```
Comment (djhoese) [inline satpy/tests/compositor_tests/test_viirs.py]: Maybe it is github's weird formatting but this looks like it needs a `self` as the first argument.

File: satpy/tests/compositor_tests/test_viirs.py [hunk 19]
```diff
@@ -223,6 +242,254 @@ def test_hncc_dnb(self):
...
+        self.assertEqual(res.attrs['platform_name'], 'Suomi-NPP')
+        self.assertEqual(res.attrs['sensor'], 'viirs')
+        self.assertEqual(res.attrs['units'], '%')
+        self.assertEqual(res.attrs['start_time'], datetime.datetime(2012, 2, 25, 18, 1, 24, 570942))
+        self.assertEqual(res.attrs['end_time'], datetime.datetime(2012, 2, 25, 18, 11, 21, 175760))
+        self.assertEqual(res.attrs['area'], area)
+        self.assertEqual(res.attrs['ancillary_variables'], [])
+        data = res.values
+        if abs(np.mean(data) - 40.7578684169142) >= 1e-10:
+            raise AssertionError('{} is not within {} of {}'.format(np.mean(data), 1e-10, 40.7578684169142))
```
Comment (djhoese) [inline satpy/tests/compositor_tests/test_viirs.py]: Replace with `self.assertX`

File: satpy/tests/compositor_tests/test_viirs.py [hunk 20]
```diff
@@ -223,6 +242,254 @@ def test_hncc_dnb(self):
...
+        np.testing.assert_allclose(unique, [25.20341702519979, 52.38819447051263, 75.79089653845898])
+
+    def test_reflectance_corrector_modis(self):
+        import xarray as xr
+        import dask.array as da
+        import numpy as np
+        import datetime
+        from satpy.composites.viirs import ReflectanceCorrector
+        from satpy import DatasetID
+        sataa_did = DatasetID(name='satellite_azimuth_angle', wavelength=None, resolution=None, polarization=None,
```
Comment (djhoese) [inline satpy/tests/compositor_tests/test_viirs.py]: None parameters can be removed

File: satpy/tests/compositor_tests/test_viirs.py [hunk 21]
```diff
@@ -223,6 +242,254 @@ def test_hncc_dnb(self):
...
+            modifiers=('sunz_corrected', 'rayleigh_corrected_crefl'), sensor='modis')
+        self.assertEqual(ref_cor.attrs['modifiers'], ('sunz_corrected', 'rayleigh_corrected_crefl'))
+        self.assertEqual(ref_cor.attrs['calibration'], 'reflectance')
+        self.assertEqual(ref_cor.attrs['wavelength'], (0.62, 0.645, 0.67))
+        self.assertEqual(ref_cor.attrs['name'], '1')
+        self.assertEqual(ref_cor.attrs['resolution'], 250)
+        self.assertEqual(ref_cor.attrs['sensor'], 'modis')
+        self.assertEqual(ref_cor.attrs['prerequisites'], [])
+        self.assertEqual(ref_cor.attrs['optional_prerequisites'], [
+            DatasetID(name='satellite_azimuth_angle', wavelength=None, resolution=None, polarization=None,
```
Comment (djhoese) [inline satpy/tests/compositor_tests/test_viirs.py]: Same here
